### ISSUE TITLE
New lint suggestion: Avoid inherent methods on generic smart pointers

### ISSUE THREAD

**Issue by madsmtm** [rust-lang/rust-clippy#11820]
### What it does

Generic smart pointers like `Box<T>`, `Rc<T>` and so on [are recommended](https://doc.rust-lang.org/nightly/std/ops/trait.Deref.html#when-to-implement-deref-or-derefmut) to use associated functions instead of inherent methods to avoid compatibility hazards with `Deref`. As an example, see [`Box::leak`](https://doc.rust-lang.org/nightly/std/boxed/struct.Box.html#method.leak). This is also [documented in the Rust API Guidelines](https://rust-lang.github.io/api-guidelines/predictability.html#smart-pointers-do-not-add-inherent-methods-c-smart-ptr).

This lint warns on every inherent method on types that implement `Deref` in a fully generic fashion.

(In particular, it should activate when `type Target = T`, but not when `type Target = u8` or `type Target = [T]`).

### Advantage

- Avoid confusing overlapping methods.
- Prevent SemVer issues with adding new inherent methods to smart pointers (which would potentially be shadowing user-defined methods).

### Drawbacks

- Using the new code is slightly more cumbersome for the user.
- Changing the code as the lint suggests would be a breaking change for an already released crate.

### Example

```rust
use std::ops::Deref;

pub struct MySmartPointer<T>(pub T);

impl<T> Deref for MySmartPointer<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T> MySmartPointer<T> {
    pub fn foo(&self) {
        unimplemented!()
    }
}
```

Could be written as:

```rust
// ... snip

impl<T> MySmartPointer<T> {
    pub fn foo(this: &Self) {
        unimplemented!()
    }
}
```

Or, in the case where the user has read and understood the drawbacks, be explicitly allowed:

```rust
// ... snip

impl<T> MySmartPointer<T> {
    #[allow(clippy::bikeshed_lint_name)]
    pub fn foo(&self) {
        unimplemented!()
    }
}
```

**Comment by Centri3** [comment]
Some (sometimes unjustifiably) use `Deref` for newtypes, [even large projects like bevy](https://docs.rs/bevy/latest/bevy/prelude/derive.Deref.html). This would afaik lint if it's generic, as an example, `struct Meters<T: num::Float>(pub T)`

In the case fully generic means no trait bounds, yeah this would work. Even if it's a newtype/marker type that would still be prone to issues
